src/object/srv_obj_remote.c

```diff
@@ -0,0 +1,299 @@
+/**
+ * (C) Copyright 2019 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+/**
+ * object server operations
+ *
+ * This file contains the object server remote object api.
+ */
+#define D_LOGFAC	DD_FAC(object)
+
+#include <uuid/uuid.h>
+
+#include <abt.h>
+#include <daos/rpc.h>
+#include <daos_srv/pool.h>
+#include <daos_srv/rebuild.h>
+#include <daos_srv/container.h>
+#include <daos_srv/vos.h>
+#include <daos_srv/bio.h>
+#include <daos_srv/daos_server.h>
+#include <daos_srv/dtx_srv.h>
+#include "obj_rpc.h"
+#include "obj_internal.h"
+
+struct obj_remote_cb_arg {
+	dtx_exec_shard_comp_cb_t	comp_cb;
+	void				*cb_arg;
+	crt_rpc_t			*parent_req;
+	struct dtx_exec_shard_arg	*shard_arg;
+};
+
+static void
+shard_update_req_cb(const struct crt_cb_info *cb_info)
+{
+	crt_rpc_t			*req = cb_info->cci_rpc;
+	struct obj_remote_cb_arg	*arg = cb_info->cci_arg;
+	struct dtx_exec_shard_arg	*shard_arg = arg->shard_arg;
+	crt_rpc_t			*parent_req = arg->parent_req;
+	struct obj_rw_out		*orwo = crt_reply_get(req);
+	struct obj_rw_in		*orw_parent = crt_req_get(parent_req);
+	int				rc = cb_info->cci_rc;
+	int				rc1 = 0;
+
+	if (orw_parent->orw_map_ver < orwo->orw_map_version) {
+		D_DEBUG(DB_IO, DF_UOID": map_ver stale (%d < %d).\n",
+			DP_UOID(orw_parent->orw_oid), orw_parent->orw_map_ver,
+			orwo->orw_map_version);
+		rc1 = -DER_STALE;
+	} else {
+		rc1 = orwo->orw_ret;
+		if (rc1 == -DER_INPROGRESS) {
+			daos_dti_copy(&shard_arg->exec_dce.dce_xid,
+				      &orwo->orw_dti_conflict);
+			shard_arg->exec_dce.dce_dkey = orwo->orw_dkey_conflict;
+		}
+	}
+
+	if (rc >= 0)
+		rc = rc1;
+
+	if (arg->comp_cb)
+		arg->comp_cb(arg->cb_arg, rc);
+
+	crt_req_decref(parent_req);
+	D_FREE_PTR(arg);
+}
+
+/* Execute update on the remote target */
+int
+ds_obj_remote_update(struct dtx_handle *dth, void *data, int idx,
+		     dtx_exec_shard_comp_cb_t comp_cb, void *cb_arg)
+{
+	struct ds_obj_exec_arg		*obj_exec_arg = data;
```

Comment:
data is obfuscated here...why not just pass in a struct with two pointers or two pointers for exec_arg and shared_args?   They can point into a contiguous buffer but what is the benefit of passing in  a void *?   It could also be something like
struct exec_data {
     struct ds_obj_exec_arg exec_arg;
     struct dtx_exec_shard_arg shard_ags[0];     
};

Reply:
ah, ok